tracee-ebpf/tracee/tracee.bpf.c

```diff
@@ -2874,20 +3024,262 @@ int BPF_KPROBE(trace_security_socket_bind)
+        if (!should_trace()) {
+            return 0;
+        }
+    }
+
+    u16 family = get_sock_family(sk);
+
+    if (family == AF_INET) {
+        net_conn_v4_t net_details = {};
+        get_network_details_from_sock_v4(sk, &net_details, 0);
+        get_local_net_id_from_network_details_v4(sk, &connect_id, &net_details, family);
+
+        debug_event.local_addr.s6_addr32[3] = net_details.local_address;
+        debug_event.local_addr.s6_addr16[5] = 0xffff;
+        debug_event.local_port = net_details.local_port;
+        debug_event.remote_addr.s6_addr32[3] = net_details.remote_address;
+        debug_event.remote_addr.s6_addr16[5] = 0xffff;
+        debug_event.remote_port = net_details.remote_port;
+    } else if (family == AF_INET6) {
+        net_conn_v6_t net_details = {};
+        get_network_details_from_sock_v6(sk, &net_details, 0);
+        get_local_net_id_from_network_details_v6(sk, &connect_id, &net_details, family);
+
+        debug_event.local_addr = net_details.local_address;
+        debug_event.local_port = net_details.local_port;
+        debug_event.remote_addr = net_details.remote_address;
+        debug_event.remote_port = net_details.remote_port;
+    } else {
+        return 0;
+    }
+
+    switch (new_state) {
+    case TCP_SYN_SENT:
+        // When we have an outgoing network connection to a remote server, we get 'TCP_ESTABLISHED' with a context
+        // which is different from the client's context.
+        // We use 'TCP_SYN_SENT' state change, which we observed to always happen in the correct context,
+        // and save the socket in sock_ctx_map so we can avoid performing the should_trace() check
+        // Note that in this state, the port equals to 0, so we don't update the network_map here
+        net_ctx_ext.host_tid = bpf_get_current_pid_tgid();
+        bpf_get_current_comm(&net_ctx_ext.comm, sizeof(net_ctx_ext.comm));
+        net_ctx_ext.local_port = connect_id.port;
+        bpf_map_update_elem(&sock_ctx_map, &sk, &net_ctx_ext, BPF_ANY);
+        break;
+    case TCP_LISTEN:
+    case TCP_ESTABLISHED:
+        // Ideally, we would update the network map only in 'TCP_ESTABLISHED' state.
+        // However, in the case of a local server program and a local client program, which communicate via the 'lo' interface,
+        // the change to the 'TCP_ESTABLISHED' state of the server's socket doesn't happen in the (process) context of the server.
+        // To update the network_map correctly in that case, we use the 'TCP_LISTEN' state.
+        if (connect_id.port) {
+            if (!sock_ctx_p) {
+                net_ctx_ext.host_tid = bpf_get_current_pid_tgid();
+                bpf_get_current_comm(&net_ctx_ext.comm, sizeof(net_ctx_ext.comm));
+                net_ctx_ext.local_port = connect_id.port;
+                bpf_map_update_elem(&sock_ctx_map, &sk, &net_ctx_ext, BPF_ANY);
+                bpf_map_update_elem(&network_map, &connect_id, &net_ctx_ext, BPF_ANY);
```

Comment:
shouldn't the value be of type net_ctx_t?

Reply:
No. This is by design to save some cycles. Notice that net_ctx is a "sub-struct" of net_ctx_ext. By updating the map with the net_ctx_ext, only the relevant fields will be saved